include/thread_pool/slotted_bag.hpp

```diff
@@ -0,0 +1,174 @@
+#pragma once
+
+#include <thread_pool/mpmc_bounded_queue.hpp>
+
+#include <atomic>
+#include <limits>
+
+namespace tp
+{
+
+/**
+* @brief The Slotted Bag is a lockless container which tracks membership by id. 
+* The bag is divided into slots, each with a unique id, and each slot can be full or empty.
+* The bag supports filling and emptying slots, as well as the unordered emptying of any full slot. 
+* The bag supports multiple consumers, and a single producer per slot.
+*/
+template <template<typename> class Queue>
+class SlottedBag
+{
+    /**
+    * @brief The implementation of a single slot of the bag. Maintains state and id.
+    * @details State Machine:
+    *
+    *             +------ TryRemoveAny -------+-------------------------------+ 
+    *             |                           |                               | 
+    *             v                           |                               | 
+    *       +-----------+              +-------------+                 +---------------+           
+    *       | NotQueued | -- add() --> | QueuedValid | -- remove() --> | QueuedInvalid |
+    *       +-----------+              +-------------+                 +---------------+
+    *                                         ^                               |
+    *                                         |                               |
+    *                                         +------------- add() -----------+
+    */
+    struct Slot
+    {
+        enum class State
+        {
+            NotQueued,
+            QueuedValid,
+            QueuedInvalid,
+        };
+
+        std::atomic<State> state;
+        size_t id;
+
+        Slot()
+            : state(State::NotQueued)
+            , id(0)
+        {
+        }
+
+        Slot(const Slot&) = delete;
+        Slot& operator=(const Slot&) = delete;
+
+        Slot(Slot&& rhs) = default;
+        Slot& operator=(Slot&& rhs) = default;
+    };
+
+public:
+
+    /**
+    * @brief SlottedBag Constructor.
+    * @param size Power of 2 number - queue length.
+    * @throws std::invalid_argument if size is not a power of 2.
+    */
+    explicit SlottedBag(size_t size);
+
+    SlottedBag(SlottedBag& rhs) = delete;
+    SlottedBag& operator=(SlottedBag& rhs) = delete;
+
+    SlottedBag(SlottedBag&& rhs) = default;
+    SlottedBag& operator=(SlottedBag&& rhs) = default;
+
+    /**
+    * @brief SlottedBag destructor.
+    */
+    virtual ~SlottedBag() = default;
```

Comment:
make class final and remove virtual methods?

Reply:
Fixed!